src/www/model/server.ts

```diff
@@ -40,10 +43,21 @@ export interface Server {
   checkReachable(): Promise<boolean>;
 }
 
-export type ServerConfig = ShadowsocksConfig;
+
+export interface ServerConfig {
+  // Static proxy configuration.
+  readonly proxy?: ShadowsocksConfig;
+
+  // Dynamic proxy configuration retrieval source.
+  source?: ShadowsocksConfigSource;
+
+  // User-given name.
+  name?: string;
+}
 
 export interface ServerRepository {
   add(serverConfig: ServerConfig): void;
+  update(serverId: string, newConfig: ServerConfig): void;
```

Comment:
This could be a method on the ConfigSource instead. On permanent redirection + successful connection, we can call `configSource.setAccessKey(newAccessKey)`, and the implementation takes care of persisting that.